src/plugins/formulas/engine/register.js

```diff
@@ -0,0 +1,224 @@
+import staticRegister from '../../../utils/staticRegister';
+import { isUndefined } from '../../../helpers/mixed';
+import { warn } from '../../../helpers/console';
+import { PLUGIN_KEY } from '../formulas';
+import { DEFAULT_LICENSE_KEY, getEngineSettingsWithDefaultsAndOverrides } from './settings';
+
+const ENGINE_KEY = 'engine';
+
+/**
+ * Setups the engine instance. It either creates a new (possibly shared) engine instance, or attaches
+ * the plugin to an already-existing instance.
+ *
+ * @param {Handsontable} hotInstance Handsontable instance.
+ * @returns {null|object} Returns the engine instance if everything worked right and `null` otherwise.
+ */
+export function setupEngine(hotInstance) {
+  const hotSettings = hotInstance.getSettings();
+  const pluginSettings = hotSettings[PLUGIN_KEY];
+  const engineConfigItem = pluginSettings?.engine;
+
+  if (pluginSettings === true) {
+    return null;
+  }
+
+  if (isUndefined(engineConfigItem)) {
+    return null;
+  }
+
+  // `engine.hyperformula` or `engine` is the engine class
+  if (typeof engineConfigItem.hyperformula === 'function' || typeof engineConfigItem === 'function') {
+    return registerEngine(
+      engineConfigItem.hyperformula ?? engineConfigItem,
+      hotSettings,
+      hotInstance);
+
+    // `engine` is the engine instance
+  } else if (typeof engineConfigItem === 'object' && isUndefined(engineConfigItem.hyperformula)) {
+    const engineRegistry = staticRegister(PLUGIN_KEY).getItem(ENGINE_KEY);
+    const sharedEngineUsage = engineRegistry?.get(engineConfigItem);
+
+    if (sharedEngineUsage) {
+      sharedEngineUsage.push(hotInstance);
+    }
+
+    if (!engineConfigItem.getConfig().licenseKey) {
+      engineConfigItem.updateConfig({
+        licenseKey: DEFAULT_LICENSE_KEY
+      });
+    }
+
+    return engineConfigItem;
+  }
+
+  return null;
+}
+
+/**
+ * Registers the engine in the global register and attaches the needed event listeners.
+ *
+ * @param {Function} engineClass The engine class.
+ * @param {object} hotSettings The Handsontable settings.
+ * @param {Handsontable} hotInstance Handsontable instance.
+ * @returns {object} Returns the engine instance.
+ */
+export function registerEngine(engineClass, hotSettings, hotInstance) {
+  if (!staticRegister(PLUGIN_KEY).hasItem(ENGINE_KEY)) {
+    staticRegister(PLUGIN_KEY).register(ENGINE_KEY, new Map());
+  }
+
+  const pluginSettings = hotSettings[PLUGIN_KEY];
+  const engineSettings = getEngineSettingsWithDefaultsAndOverrides(hotSettings);
+  const engineRegistry = staticRegister(PLUGIN_KEY).getItem(ENGINE_KEY);
+
+  registerCustomFunctions(engineClass, pluginSettings.functions);
```

Comment:
Note: this should be replaced with https://handsontable.github.io/hyperformula/api/interfaces/configparams.html#functionplugins `functionPlugins` config options, to initialize engine in one go